rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,143 @@
+
+With the example terraform, minaform can now use the `coda-network` tool (aka carey's key service), to look for a set of keys and genesis block in google cloud for this testnet name, and if they do not exist then it will generate all of the neccessary keys for you, as well as the ledger itself. In the initial implementation, this will be handled by the already ready to use but fairly hacky scripts `generate-keys-and-ledger.sh`, but as the coda-network tool becomes more fully featured then the sed+jq steps can be removed and minaform can interract directly with coda-network. The vision for coda-network includes commands for extending existing keysets or regenerating a genesis ledger with minor changes, so that once a network is created with minaform, any future edits to the keypairs or keysets or genesis ledger can be performed directly with coda-network.
+
+Flags to the `create-network` command for `--generate-keys`, `--upload-keys`, and `--download-keys` allow you to force regeneration/re-upload of keys for easier testing and quick iteration on the key generation tooling. The default value of `true` for each of these makes it easy to share keys across O(1)  Labs employees who are collaborating on the same network (Pain Point #6).
+
+The flag `--bake-image=true` will cause the tool to also run `minaform bake-image` to build a new docker image with all of the require network configuration and s3-cached files baked-in.
+
+Two flags are availible to specify which version of the daemon to run, `--daemon-image` and `--daemon-git-sha` where daemon-git-sha calculates the image tag based on the given git tag (Pain Point #5). 
+
+Additional flags can be provided to the create-network command for changing the quantities of fish, whales, seeds, bots, etc. These quantaties are provided at this stage so that the genesis ledger can be populated with proper accounts for these services.
+
+Once the network has been defined and set up, you can always make manual changes to the terraform files to customize the new network, or just proceed to `minaform start <testnet_name>`. 
+
+#### terraform.env
+
+The configured docker image, git tag, deb package tag and the image generated by create-network (with deb package + keys + genesis proof) are stored alongside the default network configuration (prod or qa) in a `terraform.env` file and in `export TF_BLAH=blah` format so that those variables can be easily validated by the deploying engineer, and updated for each iteration of a testnet.  Google cloud storage paths to the keyfiles and ledger are also provided in this file for ease of use and visibility for anyone working with this network, and for use in `minaform get-network`.
```

Comment:
Not sure I get the use of sourcing an environment file here; especially since `terraform` supports the use of *tfvars* files meant to capture sets of configs like this natively. Also, not to mention that expressing HCL (Terraform's object/resource definition language) as environment strings is likely to be a pain to configure in certain cases and considerably convoluted and difficult to read in others - think nested maps and lists. (e.g. when expressed as variable overrides).

I can somewhat see using something like this .env file as a means to enable overrides of *tfvars* files themselves as mentioned representing environment/user/adhoc/etc default configs but it almost feels unnecessary to introduce another dependency considering *tvfars* also support overrides through and with established merge precedence orders.

Reply:
I was not aware of the official tfvars files, absolutely that's the way to go instead of a custom implementation that gets sourced in bash. In either case it allows for a script to create a simple file that terraform reads, and from there terraform does the majority of the work.